File: app/src/main/java/com/blaxsoftware/directcallwidget/image/LoadImageTask.java
/*
 * Direct Call Widget - The widget that makes contacts accessible
 * Copyright (C) 2020 Fer P. A.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

package com.blaxsoftware.directcallwidget.image;

import android.content.Context;
import android.graphics.Bitmap;
import android.net.Uri;
import android.os.AsyncTask;

import com.blaxsoftware.directcallwidget.Constants;
import com.blaxsoftware.directcallwidget.util.ImageUtils;

import java.io.File;
import java.io.FileNotFoundException;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.OutputStream;

public class LoadImageTask extends AsyncTask<Uri, Void, Bitmap> {

    private Context mContext;
    private int mReqWidth;
    private int mReqHeight;
    private OnImageLoadedListener mOnImageLoadedListener;
    private boolean mCopyBitmapToInternalDir;
    private String mOutUri;

    public LoadImageTask(Context context, int reqWidth, int reqHeight, boolean copyBitmapToInternalDir) {
        mContext = context;
        mReqWidth = reqWidth;
        mReqHeight = reqHeight;
        mCopyBitmapToInternalDir = copyBitmapToInternalDir;
    }

    @Override
    protected Bitmap doInBackground(Uri... params) {
        Uri imageUri = params[0];
        mOutUri = imageUri.toString();

        Bitmap bitmap = null;
        boolean isFileUri = ImageUtils.isFileUri(mContext, imageUri);
        String source = isFileUri ? ImageUtils
                .imageFilePath(mContext, imageUri) : imageUri.toString();
        try {
            bitmap = ImageUtils.loadSampledBitmap(mContext, source, !isFileUri,
                    mReqWidth, mReqHeight);
        } catch (FileNotFoundException e) {
        }
        if (mCopyBitmapToInternalDir && bitmap != null) {
            try {
                mOutUri = copyTo(mContext, bitmap, Constants.PICTURES_DIRECTORY);
            } catch (IOException e) {}
        }
        return bitmap;
    }

    /**
     * Copies the given bitmap to an internal directory and returns its URI.
     * @param context
     * @param bitmap
     * @param internalDirName
     * @return
     * @throws IOException
     */
    private static String copyTo(Context context, Bitmap bitmap, String internalDirName) throws IOException {
        OutputStream os = null;
        try {
            File dir = new File(context.getFilesDir(), internalDirName);
            if (!dir.exists()) {
                dir.mkdirs();
            }
            File outFile = new File(dir, String.valueOf(System.currentTimeMillis()));
            os = new FileOutputStream(outFile);
            bitmap.compress(Bitmap.CompressFormat.JPEG, 100, os);
            return outFile.toURI().toString();
        } finally {
            if (os != null) {
                os.close();
            }
        }
    }

    @Override
    protected void onPostExecute(Bitmap result) {
        if (mOnImageLoadedListener != null) {
            mOnImageLoadedListener.onImageLoaded(mOutUri, result);
        }
    }

    public void setOnImageLoadedListener(
            OnImageLoadedListener onImageLoadedListener) {
        mOnImageLoadedListener = onImageLoadedListener;
    }

    public interface OnImageLoadedListener {

        void onImageLoaded(String uri, Bitmap bitmap);
    }
}

File: app/src/main/java/com/blaxsoftware/directcallwidget/image/LoadImageTask.kt
package com.blaxsoftware.directcallwidget.image

import android.content.Context
import android.graphics.Bitmap
import android.net.Uri
import android.os.AsyncTask
import android.os.Debug
import com.blaxsoftware.directcallwidget.Constants
import com.blaxsoftware.directcallwidget.util.ImageUtils
import java.io.*

typealias OnImageLoadedListener = (String, Bitmap) -> Unit

class LoadImageTask(private val mContext: Context,
                    private val mReqWidth: Int,
                    private val mReqHeight: Int,
                    private val mCopyBitmapToInternalDir: Boolean) : AsyncTask<Uri, Void, Bitmap?>() {

    private var mOnImageLoadedListener: OnImageLoadedListener? = null
    private var mOutUri: String? = null

    override fun doInBackground(vararg uris: Uri): Bitmap? {
        if (uris.isEmpty()) {
            throw IllegalArgumentException("No image uris found")
        }
        val imageUri = uris[0]
        mOutUri = imageUri.toString()

        var bitmap: Bitmap? = null
        val isFileUri = ImageUtils.isFileUri(mContext, imageUri)
        val source = if (isFileUri) {
            ImageUtils.imageFilePath(mContext, imageUri)
        } else {
            imageUri.toString()
        }
        try {
            bitmap = ImageUtils.loadSampledBitmap(mContext, source, !isFileUri,
                    mReqWidth, mReqHeight)
        } catch (e: FileNotFoundException) {
        }

        if (mCopyBitmapToInternalDir && bitmap != null) {
            try {
                mOutUri = copyTo(mContext, bitmap, Constants.PICTURES_DIRECTORY)
            } catch (e: IOException) {
            }

        }
        return bitmap
    }

    override fun onPostExecute(result: Bitmap?) {
        if (mOutUri != null && result != null) {
            mOnImageLoadedListener?.invoke(mOutUri!!, result)
        }
    }

    /**
     * Copies the given bitmap to an internal directory and returns its URI.
     * @param context
     * *
     * @param bitmap
     * *
     * @param internalDirName
     * *
     * @return
     * *
     * @throws IOException
     */
    @Throws(IOException::class)
    private fun copyTo(context: Context, bitmap: Bitmap, internalDirName: String): String {
        var os: OutputStream? = null
        try {
            val dir = File(context.filesDir, internalDirName)
            if (!dir.exists()) {
                dir.mkdirs()
            }
            val outFile = File(dir, System.currentTimeMillis().toString())
            os = FileOutputStream(outFile)
            bitmap.compress(Bitmap.CompressFormat.JPEG, 100, os)
            return outFile.toURI().toString()
        } finally {
            if (os != null) {
                os.close()
            }
        }
    }

    fun setOnImageLoadedListener(listener: OnImageLoadedListener) {
        mOnImageLoadedListener = listener
    }
}
